base/applications/regedit/find.c

```diff
@@ -223,14 +223,28 @@ BOOL RegFindRecurse(
                                   NULL, &cb);
         if (lResult != ERROR_SUCCESS)
             goto err;
-        pb = malloc(cb);
+        pb = malloc(cb + 3); /* To avoid buffer overrun, append 3 NULs */
         if (pb == NULL)
             goto err;
         lResult = RegQueryValueExW(hSubKey, ppszNames[i], NULL, &type,
                                   pb, &cb);
         if (lResult != ERROR_SUCCESS)
             goto err;
 
+        if (type == REG_SZ || type == REG_EXPAND_SZ || type == REG_MULTI_SZ)
+        {
+            /* Test for odd (not even) number of bytes returned */
+            if (cb & 1)
+                DPRINTF("Byte Count of (%d) for wide string is ODD (Not even)\n", cb);
+        }
+
+        /* To avoid buffer overrun, append 3 NUL bytes.
+           NOTE: cb can be an odd number although UNICODE_NULL is two bytes.
```

Comment:
I tested your suggestion by going to 'master' and adding your two lines at line #234. This was right after the RegQueryValueEx and its checking for returned value error. This produced a very quick crash with the log as follows:
```
Unhandled exception
ExceptionCode:    c0000005
Faulting Address: 20230A38
CS:EIP 1b:7c93f73a
DS 23 ES 23 FS 3b GS 0
EAX: 00230a3c   EBX: 0024e218   ECX: ffffffff
EDX: 07ffffff   EBP: 0022dc38   ESI: 00230000   ESP: 0022dc38
EDI: ffffffff   EFLAGS: 00010a07

*** Assertion failed: FreeEntry->PreviousSize == CurrentEntry->Size
***   Source File: ../sdk/lib/rtl/heap.c, line 1166

Break repeatedly, break Once, Ignore, terminate Process or terminate Thread (boipt)? 
kdb:> o
Execute '.cxr 0022CD60' to dump context
traywnd.cpp:3259: Unexpected failure (m_TrayBandSite->ProcessMessage(m_hWnd, uMsg, wParam, lParam, &Ret))=80004005.
fixme:(dll/win32/imm32/ctf.c:77) (04090409)
aclmulti.cpp:62: Unexpected failure (punk->QueryInterface(IID_PPV_ARG(IACList, &fObjects[fObjectCount].pACL)))=80004002.
err:(dll/win32/shlwapi/autocomp.cpp:51) punk->QueryInterface failed: 0x80004002
fixme:(dll/win32/shlwapi/url.c:1977) (L"regedit" 4): stub
fixme:(dll/win32/shell32/shlexec.cpp:1926) flags ignored: 0x00000100
fixme:(dll/win32/imm32/ctf.c:77) (04090409)
Unhandled exception
ExceptionCode:    c0000005
Faulting Address: 20230A38
CS:EIP 1b:7c93f73a
DS 23 ES 23 FS 3b GS 0
EAX: 00230a3c   EBX: 0024e218   ECX: ffffffff
EDX: 07ffffff   EBP: 0022dc38   ESI: 00230000   ESP: 0022dc38
EDI: ffffffff   EFLAGS: 00010a07

*** Assertion failed: FreeEntry->PreviousSize == CurrentEntry->Size
***   Source File: ../sdk/lib/rtl/heap.c, line 1166

Break repeatedly, break Once, Ignore, terminate Process or terminate Thread (boipt)? 
kdb:> O
Execute '.cxr 0022CD60' to dump context
```
Suggestions? Thanks for your help on this.

Reply:
> PWCHAR ValueBuffer = (PWCHAR)pb;
> ValueBuffer[cb / sizeof(WCHAR)] = UNICODE_NULL;

This broke the original data. I would reject it.

Reply:
Jesus. When a unicode string has an odd size in bytes, there is a major problem. This doesn't just happen. It means there is a bug somewhere. This whole "let's ignore the bug and just pretend it's all good" thing is not the way to go. Can we please stop adding hacks and fix the shit where it is broken? Thanks.

Reply:
I unresolved this conversation with the aim to make @tkreuzer comment more visible in the future.
Please leave it in unresolved state.

Reply:
In any case, a registry viewer/editor should not crash when it encounters malformed data.

Reply:
The data comes from the registry. If it is malformed, there is a either a bug in the kernel or - and that is more likely - a bug in some other regedit code. Either way, someone needs to investigate it instead of blindly assuming that random crashes on invalid data are simply something "normal".

Reply:
https://github.com/reactos/reactos/pull/5150/files#diff-e6020d9b127f328f907c818bb9c674c07e03b834649b131637b4c9feb607bc46R103

`DebugBreak();` will be added for odd size.

Reply:
Please don't add extra code for this. Just handle it safely without crashing. If we want to find bugs in places that write to the registry, let's have a separate scanner tool.
My suggestion from the top of this thread still stands. The only reason it crashed is because the `malloc` call needs space for the extra null terminator, just like it does with the weird 3-byte solution.